File: 547-number-of-provinces/547-number-of-provinces.cpp
class Solution
{
    public:
        void dfs(int src, vector<vector < int>> &gp, vector< bool > &vis)
        {
            vis[src] = true;
            for (int desti: gp[src])
            {
                if (vis[desti] == false)
                {
                    dfs(desti, gp, vis);
                }
            }
        }

        int findCircleNum(vector<vector < int>> &isConnected)
        {
            int n = isConnected.size();
            vector<vector < int>> gp(n);
            for (int i = 0; i < n; i++)
            {
                for (int j = 0; j < n; j++)
                {
                    if (isConnected[i][j] == 1)
                    {
                        gp[i].push_back(j);
                        gp[j].push_back(i);
                    }
                }
            }

            int numOfProvinces = 0;
            vector<bool> vis(n, false);
            for (int i = 0; i < n; i++)
            {
                if (vis[i] == false)
                {
                    dfs(i, gp, vis);
                    numOfProvinces++;
                }
            }

            return numOfProvinces;
        }
};

File: 547-number-of-provinces/547-number-of-provinces.java
class Solution {
    private void dfs(int x, int[][] isConnected, boolean[] vis) {
        vis[x] = true;
        for(int col = 0; col < isConnected[x].length; col++) {
            if(isConnected[x][col] == 1 && vis[col] == false) {
                dfs(col, isConnected, vis);
            }
        }
    }
    
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        
        int numOfProvinces = 0;
        
        boolean[] vis = new boolean[n];
        for(int i = 0; i < n; i++) {
            if(vis[i] == false) {
                numOfProvinces += 1;
                dfs(i, isConnected, vis);
            }
        }
        
        return numOfProvinces;
    }
}